File: cpp/11.cpp
#include <iostream>
#include <vector>
#include <algorithm>
#include <iterator>
#include <utility>
#include <map>
#include <string>
#include <unordered_map>
#include <climits>
using namespace std;

template<typename T>
void prints(vector<T> &v){
  cout << "[";
  for(uint i=0; i<v.size()-1; ++i)
    cout << v[i] << ", ";
  if(v.size()) cout << v[v.size()-1];
  cout << "]\n";
}

class Solution {
public:
  int maxArea(vector<int> &height) {
    int n = height.size();
    int ans = 0, l = 0, r = n-1;
    while(l < r){
      int v = min(height[l], height[r]) * (r-l);
      ans = max(ans, v);
      if(height[l] <= height[r])
        ++l;
      else
        --r;
    }
    return ans;
  }
};


int main(){
  auto sol = Solution();
}

File: python/11.py
class Solution(object):
    def maxArea(self, height):
        """
        :type height: List[int]
        :rtype: int
        """
        l, r = 0, len(height)-1
        ans = 0
        while l < r:
            ans = max(ans, min(height[l], height[r]) * (r-l))
            if height[l] <= height[r]:
                h = height[l]
                while l < r and height[l] <= h:
                    l += 1
            else:
                h = height[r]
                while l < r and height[r] <= h:
                    r -= 1
        return ans
